src/app/clusters/ota-requestor/OTARequestor.cpp

```diff
@@ -0,0 +1,389 @@
+ */
+#include "OTARequestor.h"
+
+#include <app/server/Server.h>
+#include <app/util/util.h>
+#include <lib/core/CHIPEncoding.h>
+#include <platform/CHIPDeviceLayer.h>
+
+#include "BDXDownloader.h"
+
+#include <controller/CHIPDeviceControllerFactory.h>
+#include <controller/CommissioneeDeviceProxy.h>
+#include <controller/ExampleOperationalCredentialsIssuer.h>
+
+#include <zap-generated/CHIPClientCallbacks.h>
+#include <zap-generated/CHIPClusters.h>
+
+using chip::ByteSpan;
+using chip::CASESessionManager;
+using chip::CASESessionManagerConfig;
+using chip::CharSpan;
+using chip::DeviceProxy;
+using chip::EndpointId;
+using chip::FabricIndex;
+using chip::FabricInfo;
+using chip::NodeId;
+using chip::OnDeviceConnected;
+using chip::OnDeviceConnectionFailure;
+using chip::PeerId;
+using chip::Server;
+using chip::VendorId;
+using chip::bdx::TransferSession;
+using chip::Callback::Callback;
+using chip::System::Layer;
+using chip::Transport::PeerAddress;
+// using namespace chip::ArgParser;
+using namespace chip::Messaging;
+using namespace chip::app::Clusters::OtaSoftwareUpdateProvider::Commands;
+using chip::Inet::IPAddress;
+
+// Global instance of the OTARequestorInterface.
+OTARequestorInterface * globalOTARequestorInstance = nullptr;
+
+constexpr uint32_t kImmediateStartDelayMs = 1; // Start the timer with this value when starting OTA "immediately"
+
+// Callbacks for connection management
+void OnConnected(void * context, chip::OperationalDeviceProxy * deviceProxy);
+Callback<OnDeviceConnected> mOnConnectedCallback(OnConnected, nullptr);
+
+void OnConnectionFailure(void * context, NodeId deviceId, CHIP_ERROR error);
+Callback<OnDeviceConnectionFailure> mOnConnectionFailureCallback(OnConnectionFailure, nullptr);
+
+void OnQueryImageResponse(void * context, const QueryImageResponse::DecodableType & response);
+void OnQueryImageFailure(void * context, EmberAfStatus status);
+
+void SetRequestorInstance(OTARequestorInterface * instance)
+{
+    globalOTARequestorInstance = instance;
+}
+
+OTARequestorInterface * GetRequestorInstance()
+{
+    return globalOTARequestorInstance;
+}
+
+void StartDelayTimerHandler(chip::System::Layer * systemLayer, void * appState)
+{
+    VerifyOrReturn(appState != nullptr);
+    static_cast<OTARequestor *>(appState)->ConnectToProvider();
+}
+
+void OnQueryImageFailure(void * context, EmberAfStatus status)
+{
+    ChipLogDetail(SoftwareUpdate, "QueryImage failure response %" PRIu8, status);
+}
+
+void OnConnectionFailure(void * context, NodeId deviceId, CHIP_ERROR error)
+{
+    ChipLogError(SoftwareUpdate, "failed to connect to 0x%" PRIX64 ": %" CHIP_ERROR_FORMAT, deviceId, error.Format());
+}
+// Finds the Requestor instance and calls the corresponding OTARequestor member function
+void OnQueryImageResponse(void * context, const QueryImageResponse::DecodableType & response)
+{
+    OTARequestor * requestorCore = static_cast<OTARequestor *>(GetRequestorInstance());
+
+    assert(requestorCore != nullptr);
```

Comment:
Should we assert ? Can we VerifyOrReturnError ?

Reply:
This can be debated but I believe we should assert here. If the OTARequestor instance is not set then something is very broken (like the Requestor was not initialized but we are still processing the Requestor cluster commands)